File: recipes/subms-health/python/src/subms_health/registry.py
"""The registry: indicators + per-indicator policy, an optional background
refresher thread that probes off the request path, and a pre-rendered cached
snapshot served by render(). Status changes emit subms-events events."""

from __future__ import annotations

import os
import socket
import threading
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Callable, Dict, List, Optional

from subms_events import DispatchMode, Event, EventBridge, EventDispatcher, EventLevel, EventListener

from .clock import Clock, SystemClock
from .component import ComponentHealth, HealthIndicator, fn_indicator
from .env_section import EnvSection
from .serialize import report_to_json
from .status import HealthStatus, http_status_for

HEALTH_STATUS_TOPIC = "subms.health.status"


class ProbeKind(Enum):
    LIVENESS = "LIVENESS"
    READINESS = "READINESS"
    STARTUP = "STARTUP"


class RefreshMode(Enum):
    SYNC = "SYNC"
    ASYNC = "ASYNC"


@dataclass
class RefreshPolicy:
    interval_ms: int = 30_000
    probe_kinds: List[ProbeKind] = field(default_factory=lambda: [ProbeKind.READINESS])
    critical: bool = True

    def with_interval_ms(self, ms: int) -> "RefreshPolicy":
        self.interval_ms = ms
        return self

    def with_kinds(self, kinds: List[ProbeKind]) -> "RefreshPolicy":
        self.probe_kinds = list(kinds)
        return self

    def set_critical(self, critical: bool) -> "RefreshPolicy":
        self.critical = critical
        return self

    def includes(self, kind: ProbeKind) -> bool:
        return kind in self.probe_kinds


@dataclass
class HealthConfig:
    mode: RefreshMode = RefreshMode.ASYNC
    tick_ms: int = 1_000
    dispatch: DispatchMode = DispatchMode.ASYNC
    stale_factor: float = 1.5

    @staticmethod
    def sync() -> "HealthConfig":
        return HealthConfig(mode=RefreshMode.SYNC, dispatch=DispatchMode.SYNC)


def _level_for(status: HealthStatus) -> EventLevel:
    if status == HealthStatus.DOWN:
        return EventLevel.ERROR
    if status in (HealthStatus.DEGRADED, HealthStatus.WARN):
        return EventLevel.WARN
    return EventLevel.INFO


def _code_for(status: HealthStatus, probe: Optional[ProbeKind]) -> int:
    if probe is None:
        return http_status_for(status)
    if probe == ProbeKind.LIVENESS:
        return 503 if status == HealthStatus.DOWN else 200
    if probe == ProbeKind.READINESS:
        return 503 if status in (HealthStatus.DEGRADED, HealthStatus.DOWN) else 200
    # STARTUP: not ready until UP/WARN
    return 200 if status in (HealthStatus.UP, HealthStatus.WARN) else 503


class ServerIndicator(HealthIndicator):
    def __init__(self) -> None:
        self._start = time.monotonic()
        self._pid = os.getpid()
        try:
            self._hostname = socket.gethostname()
        except Exception:
            self._hostname = "unknown"

    def name(self) -> str:
        return "server"

    def check(self) -> ComponentHealth:
        return (
            ComponentHealth.up()
            .with_detail("pid", self._pid)
            .with_detail("hostname", self._hostname)
            .with_detail("uptime_ms", int((time.monotonic() - self._start) * 1000))
        )


class HealthRegistry:
    def __init__(self, config: Optional[HealthConfig] = None, clock: Optional[Clock] = None) -> None:
        self._config = config or HealthConfig()
        self._clock = clock or SystemClock()
        self._lock = threading.Lock()
        self._indicators: List[tuple] = []  # (indicator, policy)
        self._cache: Dict[int, tuple] = {}  # index -> (component, refreshed_at_ms)
        self._snapshot: Optional[Dict[Optional[ProbeKind], tuple]] = None
        self._dispatcher = EventDispatcher(self._config.dispatch)
        self._emitter = self._dispatcher.handle()
        self._prev_overall: Optional[HealthStatus] = None
        self._prev_components: Dict[str, HealthStatus] = {}
        self._thread: Optional[threading.Thread] = None
        self._stop = threading.Event()

    # ---- construction helpers ----

    @staticmethod
    def with_system_sections(config: Optional[HealthConfig] = None) -> "HealthRegistry":
        r = HealthRegistry(config)
        r.register(
            ServerIndicator(),
            RefreshPolicy(interval_ms=5_000, probe_kinds=list(ProbeKind), critical=False),
        )
        deploy = (
            EnvSection("deploy")
            .prefix("KICKSTART_")
            .strip_prefix_in_key(True)
            .lowercase_keys(True)
            .redact_secrets()
        )
        r.register_section(
            deploy, RefreshPolicy(interval_ms=60_000, probe_kinds=list(ProbeKind), critical=False)
        )
        return r

    # ---- registration ----

    def register(self, indicator: HealthIndicator, policy: Optional[RefreshPolicy] = None) -> "HealthRegistry":
        with self._lock:
            self._indicators.append((indicator, policy or RefreshPolicy()))
        return self

    def register_fn(self, name: str, fn: Callable[[], ComponentHealth], policy: Optional[RefreshPolicy] = None) -> "HealthRegistry":
        return self.register(fn_indicator(name, fn), policy)

    def register_section(self, section: EnvSection, policy: Optional[RefreshPolicy] = None) -> "HealthRegistry":
        return self.register(section.into_indicator(), policy)

    def add_listener(self, listener: EventListener) -> "HealthRegistry":
        self._dispatcher.add_listener(listener)
        return self

    def add_bridge(self, bridge: EventBridge) -> "HealthRegistry":
        self._dispatcher.add_bridge(bridge)
        return self

    # ---- refresh ----

    def refresh_now(self) -> None:
        self._rebuild(force=True)

    def refresh_due(self) -> None:
        self._rebuild(force=False)

    def start(self) -> None:
        self._rebuild(force=True)
        if self._config.mode == RefreshMode.SYNC or self._thread is not None:
            return
        self._stop.clear()
        tick = max(1, self._config.tick_ms) / 1000.0

        def run() -> None:
            while not self._stop.wait(tick):
                self._rebuild(force=False)

        self._thread = threading.Thread(target=run, name="subms-health-refresh", daemon=True)
        self._thread.start()

    def stop(self) -> None:
        if self._thread is not None:
            self._stop.set()
            self._thread.join()
            self._thread = None
        self._dispatcher.stop()

    # ---- render ----

    def _ensure_snapshot(self) -> None:
        if self._snapshot is None:
            self._rebuild(force=True)

    def render(self) -> tuple:
        self._ensure_snapshot()
        return self._snapshot[None]

    def render_liveness(self) -> tuple:
        self._ensure_snapshot()
        return self._snapshot[ProbeKind.LIVENESS]

    def render_readiness(self) -> tuple:
        self._ensure_snapshot()
        return self._snapshot[ProbeKind.READINESS]

    def render_startup(self) -> tuple:
        self._ensure_snapshot()
        return self._snapshot[ProbeKind.STARTUP]

    def status(self) -> HealthStatus:
        self._ensure_snapshot()
        return self._snapshot["__status__"]

    # ---- internals ----

    def _rebuild(self, force: bool) -> None:
        now = self._clock.now_ms()
        stamp = self._clock.now_rfc3339()
        with self._lock:
            indicators = list(self._indicators)
            for i, (ind, pol) in enumerate(indicators):
                cached = self._cache.get(i)
                due = force or cached is None or (now - cached[1]) >= pol.interval_ms
                if due:
                    self._cache[i] = (ind.check(), now)

            snapshot: Dict[object, tuple] = {}
            overall_status = None
            for probe in (None, ProbeKind.LIVENESS, ProbeKind.READINESS, ProbeKind.STARTUP):
                status, code, body = self._build_view(indicators, now, stamp, probe)
                snapshot[probe] = (code, body)
                if probe is None:
                    overall_status = status
            snapshot["__status__"] = overall_status

            current: Dict[str, HealthStatus] = {}
            for i, (ind, _pol) in enumerate(indicators):
                cached = self._cache.get(i)
                current[ind.name()] = (
                    cached[0].effective_status() if cached else HealthStatus.UNKNOWN
                )
            self._snapshot = snapshot

        self._emit_changes(overall_status, current, stamp)

    def _build_view(self, indicators, now, stamp, probe):
        entries = []
        contributed = []
        for i, (ind, pol) in enumerate(indicators):
            if probe is not None and not pol.includes(probe):
                continue
            cached = self._cache.get(i)
            if cached is not None:
                comp, refreshed_at = cached
            else:
                comp = ComponentHealth.unknown().with_detail("state", "pending")
                refreshed_at = now
            eff = comp.effective_status()
            if (not pol.critical) and eff in (HealthStatus.DOWN, HealthStatus.DEGRADED):
                contrib = HealthStatus.WARN
            else:
                contrib = eff
            contributed.append(contrib)
            age_ms = max(0, now - refreshed_at)
            stale = age_ms > pol.interval_ms * self._config.stale_factor
            entries.append((ind.name(), eff.value, age_ms, stale, comp))
        status = HealthStatus.aggregate(contributed)
        entries.sort(key=lambda e: e[0])
        body = report_to_json(status.value, stamp, entries)
        return status, _code_for(status, probe), body

    def _emit_changes(self, overall: HealthStatus, current: Dict[str, HealthStatus], at: str) -> None:
        changes = []
        if self._prev_overall is not None and self._prev_overall != overall:
            changes.append(("overall", self._prev_overall, overall))
        self._prev_overall = overall
        for name, st in current.items():
            old = self._prev_components.get(name)
            if old is not None and old != st:
                changes.append((name, old, st))
        self._prev_components = dict(current)
        for scope, frm, to in changes:
            event = (
                Event.builder(HEALTH_STATUS_TOPIC)
                .level(_level_for(to))
                .at(at)
                .message(f"{scope}: {frm.value} -> {to.value}")
                .attr("scope", scope)
                .attr("from", frm.value)
                .attr("to", to.value)
                .build()
            )
            self._emitter.emit(event)

File: recipes/subms-health/rust/src/registry.rs
//! The registry: holds indicators + per-indicator refresh policy, runs a stdlib
//! background thread that probes each indicator on its own cadence (off the
//! request path), and serves a pre-rendered cached snapshot. `render` is a cache
//! read - it never probes, never touches a dependency, and stays sub-ms.

use std::collections::BTreeMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Condvar, Mutex};
use std::thread::{self, JoinHandle};
use std::time::{Duration, Instant};

use subms_events::{DispatchMode, EmitHandle, Event, EventBridge, EventDispatcher, EventListener};

use crate::clock::{Clock, SystemClock};
use crate::component::{ComponentHealth, HealthIndicator};
use crate::env_section::EnvSection;
use crate::events::{HEALTH_STATUS_TOPIC, level_for};
use crate::json::{JsonValue, push_component_map, push_json_str, push_map};
use crate::status::{HealthStatus, http_status_for};

/// Kubernetes-style probe classes. An indicator may serve several.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ProbeKind {
    Liveness,
    Readiness,
    Startup,
}

/// How the registry refreshes its snapshot.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RefreshMode {
    /// You drive refreshes yourself (call `refresh_due` / `refresh_now` from your
    /// own loop). `start()` just warms the cache once. No background thread.
    Sync,
    /// A background thread refreshes due indicators every `tick_ms`. `start()`
    /// spawns it.
    Async,
}

/// Registry-level config: refresh mode + cadence, event-dispatch mode, and the
/// staleness factor. The mirror of this struct exists in every language port.
///
/// The lowest-latency, zero-extra-thread setup is
/// `HealthConfig { mode: Sync, dispatch: Sync, .. }`: no background threads at
/// all, you drive `refresh_due()` from your own loop, and `render_arc()` serves
/// the cached bytes with no per-request allocation.
#[derive(Debug, Clone, Copy)]
pub struct HealthConfig {
    pub mode: RefreshMode,
    /// Async refresher cadence (ms). Ignored in Sync mode.
    pub tick_ms: u64,
    pub dispatch: DispatchMode,
    /// Flag a component `stale` once its age exceeds `stale_factor * interval`.
    pub stale_factor: f64,
}

impl Default for HealthConfig {
    fn default() -> Self {
        Self {
            mode: RefreshMode::Async,
            tick_ms: 1_000,
            dispatch: DispatchMode::Async,
            stale_factor: 1.5,
        }
    }
}

impl HealthConfig {
    pub fn new() -> Self {
        Self::default()
    }
    /// No background threads: sync refresh + sync dispatch. The low-latency path.
    pub fn sync() -> Self {
        Self {
            mode: RefreshMode::Sync,
            dispatch: DispatchMode::Sync,
            ..Self::default()
        }
    }
    pub fn async_every_ms(tick_ms: u64) -> Self {
        Self {
            mode: RefreshMode::Async,
            tick_ms,
            ..Self::default()
        }
    }
    pub fn refresh_mode(mut self, mode: RefreshMode) -> Self {
        self.mode = mode;
        self
    }
    pub fn dispatch_mode(mut self, dispatch: DispatchMode) -> Self {
        self.dispatch = dispatch;
        self
    }
    pub fn tick_ms(mut self, tick_ms: u64) -> Self {
        self.tick_ms = tick_ms;
        self
    }
    pub fn with_stale_factor(mut self, factor: f64) -> Self {
        self.stale_factor = factor;
        self
    }
}

/// Per-indicator policy: how often to refresh, which probes it serves, and
/// whether a failure is critical (a non-critical failure is demoted to Warn so
/// the overall report stays serving instead of failing).
#[derive(Debug, Clone)]
pub struct RefreshPolicy {
    pub interval_ms: u64,
    pub probe_kinds: Vec<ProbeKind>,
    pub critical: bool,
}

impl Default for RefreshPolicy {
    fn default() -> Self {
        Self {
            interval_ms: 30_000,
            probe_kinds: vec![ProbeKind::Readiness],
            critical: true,
        }
    }
}

impl RefreshPolicy {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn with_interval_ms(mut self, ms: u64) -> Self {
        self.interval_ms = ms;
        self
    }
    pub fn with_kinds(mut self, kinds: &[ProbeKind]) -> Self {
        self.probe_kinds = kinds.to_vec();
        self
    }
    pub fn critical(mut self, critical: bool) -> Self {
        self.critical = critical;
        self
    }
    fn includes(&self, kind: ProbeKind) -> bool {
        self.probe_kinds.contains(&kind)
    }
}

struct Registered {
    indicator: Arc<dyn HealthIndicator>,
    policy: RefreshPolicy,
}

struct Cached {
    component: ComponentHealth,
    refreshed_at_ms: u64,
}

/// One rendered view (a probe subset): status, http code, and the JSON bytes.
/// `json` is an `Arc<str>` so `render_arc` can hand it out with no copy.
#[derive(Clone)]
struct View {
    status: HealthStatus,
    code: u16,
    json: Arc<str>,
}

struct Snapshot {
    all: View,
    live: View,
    ready: View,
    started: View,
}

struct Inner {
    indicators: Mutex<Vec<Registered>>,
    cache: Mutex<Vec<Option<Cached>>>,
    snapshot: Mutex<Option<Arc<Snapshot>>>,
    clock: Arc<dyn Clock>,
    stale_factor: f64,
    emitter: EmitHandle,
    prev_overall: Mutex<Option<HealthStatus>>,
    prev_components: Mutex<BTreeMap<String, HealthStatus>>,
}

struct Refresher {
    stop: Arc<AtomicBool>,
    wake: Arc<Condvar>,
    lock: Arc<Mutex<()>>,
    handle: Option<JoinHandle<()>>,
}

/// A health registry. Build it, register indicators, optionally `start()` the
/// background refresher, then serve `render*()` from your HTTP handler.
pub struct HealthRegistry {
    inner: Arc<Inner>,
    refresher: Option<Refresher>,
    dispatcher: EventDispatcher,
    mode: RefreshMode,
    tick_ms: u64,
}

impl HealthRegistry {
    /// Empty registry with default config on the system clock.
    pub fn new() -> Self {
        Self::with_config(HealthConfig::default())
    }

    /// Build from an explicit config on the system clock.
    pub fn with_config(config: HealthConfig) -> Self {
        Self::with_config_and_clock(config, Arc::new(SystemClock))
    }

    /// Default config with a custom clock (used by tests / fixtures).
    pub fn with_clock(clock: Arc<dyn Clock>) -> Self {
        Self::with_config_and_clock(HealthConfig::default(), clock)
    }

    /// Build from an explicit config + clock.
    pub fn with_config_and_clock(config: HealthConfig, clock: Arc<dyn Clock>) -> Self {
        let dispatcher = EventDispatcher::new(config.dispatch);
        let emitter = dispatcher.handle();
        Self {
            inner: Arc::new(Inner {
                indicators: Mutex::new(Vec::new()),
                cache: Mutex::new(Vec::new()),
                snapshot: Mutex::new(None),
                clock,
                stale_factor: config.stale_factor,
                emitter,
                prev_overall: Mutex::new(None),
                prev_components: Mutex::new(BTreeMap::new()),
            }),
            refresher: None,
            dispatcher,
            mode: config.mode,
            tick_ms: config.tick_ms,
        }
    }

    /// Flag a component `stale` once its age exceeds `factor * interval`. Below 1
    /// is an early warning; above 1 an overdue alarm. Default 1.5. Must be set
    /// before any indicator is registered (it lives in the shared inner state).
    pub fn with_stale_factor(mut self, factor: f64) -> Self {
        if let Some(inner) = Arc::get_mut(&mut self.inner) {
            inner.stale_factor = factor;
        }
        self
    }

    /// A registry pre-loaded with a `server` indicator (pid/host/uptime) and a
    /// `deploy` env section (every `KICKSTART_*` var, secrets masked).
    pub fn with_system_sections() -> Self {
        let mut r = Self::new();
        r.register(
            Arc::new(ServerIndicator::new()),
            RefreshPolicy::new()
                .with_interval_ms(5_000)
                .with_kinds(&[
                    ProbeKind::Liveness,
                    ProbeKind::Readiness,
                    ProbeKind::Startup,
                ])
                .critical(false),
        );
        let deploy = EnvSection::new("deploy")
            .prefix("KICKSTART_")
            .strip_prefix_in_key(true)
            .lowercase_keys(true)
            .redact_secrets();
        r.register_section(
            deploy,
            RefreshPolicy::new()
                .with_interval_ms(60_000)
                .with_kinds(&[
                    ProbeKind::Liveness,
                    ProbeKind::Readiness,
                    ProbeKind::Startup,
                ])
                .critical(false),
        );
        r
    }

    /// Register an indicator with a refresh policy.
    pub fn register(
        &mut self,
        indicator: Arc<dyn HealthIndicator>,
        policy: RefreshPolicy,
    ) -> &mut Self {
        self.inner
            .indicators
            .lock()
            .unwrap()
            .push(Registered { indicator, policy });
        self
    }

    /// Register a closure indicator.
    pub fn register_fn<F>(&mut self, name: &str, policy: RefreshPolicy, f: F) -> &mut Self
    where
        F: Fn() -> ComponentHealth + Send + Sync + 'static,
    {
        self.register(Arc::new(crate::component::indicator(name, f)), policy)
    }

    /// Register an env section bound to the real process environment.
    pub fn register_section(&mut self, section: EnvSection, policy: RefreshPolicy) -> &mut Self {
        self.register(Arc::new(section.into_system_indicator()), policy)
    }

    /// Subscribe to status-change events (overall + per-component transitions,
    /// emitted as `subms-events` events on topic `subms.health.status`). Delivery
    /// honors the configured dispatch mode (async off-thread by default), so a
    /// slow listener never stalls a probe.
    pub fn add_listener(&mut self, listener: Arc<dyn EventListener>) -> &mut Self {
        self.dispatcher.add_listener(listener);
        self
    }

    /// Attach an event bridge (e.g. `subms-otel`'s `OtelEventBridge`) that
    /// forwards status-change events to an external system.
    pub fn add_bridge(&mut self, bridge: Arc<dyn EventBridge>) -> &mut Self {
        self.dispatcher.add_bridge(bridge);
        self
    }

    /// Force-probe every indicator now and rebuild the cached snapshot.
    pub fn refresh_now(&self) {
        rebuild(&self.inner, true);
    }

    /// Probe only the indicators whose interval has elapsed, then rebuild the
    /// snapshot. Call this from your own event loop instead of `start()` if you
    /// already own a scheduler.
    pub fn refresh_due(&self) {
        rebuild(&self.inner, false);
    }

    /// Warm the cache, and in Async mode spawn the background refresher (a
    /// daemon-style thread that wakes every `tick_ms`, probes due indicators, and
    /// swaps in a fresh snapshot). In Sync mode this only warms the cache - you
    /// drive `refresh_due()` yourself. Honors the configured `RefreshMode`.
    pub fn start(&mut self) {
        rebuild(&self.inner, true);
        if self.mode == RefreshMode::Sync || self.refresher.is_some() {
            return;
        }
        let tick_ms = self.tick_ms;
        let stop = Arc::new(AtomicBool::new(false));
        let wake = Arc::new(Condvar::new());
        let lock = Arc::new(Mutex::new(()));
        let inner = Arc::clone(&self.inner);
        let t_stop = Arc::clone(&stop);
        let t_wake = Arc::clone(&wake);
        let t_lock = Arc::clone(&lock);
        let handle = thread::Builder::new()
            .name("subms-health-refresh".to_string())
            .spawn(move || {
                let tick = Duration::from_millis(tick_ms.max(1));
                while !t_stop.load(Ordering::Acquire) {
                    let guard = t_lock.lock().unwrap();
                    let _ = t_wake.wait_timeout(guard, tick);
                    if t_stop.load(Ordering::Acquire) {
                        break;
                    }
                    rebuild(&inner, false);
                }
            })
            .expect("spawn refresher");
        self.refresher = Some(Refresher {
            stop,
            wake,
            lock,
            handle: Some(handle),
        });
    }

    /// Stop the background refresher and the event dispatcher, joining both.
    pub fn stop(&mut self) {
        if let Some(mut r) = self.refresher.take() {
            r.stop.store(true, Ordering::Release);
            {
                let _g = r.lock.lock().unwrap();
                r.wake.notify_all();
            }
            if let Some(h) = r.handle.take() {
                let _ = h.join();
            }
        }
        self.dispatcher.stop();
    }

    fn view_arc(&self, pick: fn(&Snapshot) -> &View) -> (u16, Arc<str>) {
        let need_refresh = self.inner.snapshot.lock().unwrap().is_none();
        if need_refresh {
            rebuild(&self.inner, true);
        }
        let snap = self.inner.snapshot.lock().unwrap();
        let v = pick(snap.as_ref().expect("snapshot present"));
        (v.code, Arc::clone(&v.json))
    }

    /// The full `/health` document: `(http_code, json)`. Cache read, never probes.
    pub fn render(&self) -> (u16, String) {
        let (code, json) = self.view_arc(|s| &s.all);
        (code, json.to_string())
    }
    /// Zero-copy variant of [`render`]: clones an `Arc<str>` instead of allocating
    /// a fresh `String`. The lowest-latency serve path.
    pub fn render_arc(&self) -> (u16, Arc<str>) {
        self.view_arc(|s| &s.all)
    }
    /// `/health/live` - liveness probes only.
    pub fn render_liveness(&self) -> (u16, String) {
        let (code, json) = self.view_arc(|s| &s.live);
        (code, json.to_string())
    }
    /// `/health/ready` - readiness probes only.
    pub fn render_readiness(&self) -> (u16, String) {
        let (code, json) = self.view_arc(|s| &s.ready);
        (code, json.to_string())
    }
    /// `/health/started` - startup probes only.
    pub fn render_startup(&self) -> (u16, String) {
        let (code, json) = self.view_arc(|s| &s.started);
        (code, json.to_string())
    }

    /// Overall status of the full document.
    pub fn status(&self) -> HealthStatus {
        let need_refresh = self.inner.snapshot.lock().unwrap().is_none();
        if need_refresh {
            rebuild(&self.inner, true);
        }
        self.inner
            .snapshot
            .lock()
            .unwrap()
            .as_ref()
            .unwrap()
            .all
            .status
    }
}

impl Default for HealthRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for HealthRegistry {
    fn drop(&mut self) {
        self.stop();
    }
}

fn rebuild(inner: &Inner, force: bool) {
    let now = inner.clock.now_ms();
    let stamp = inner.clock.now_rfc3339();
    let inds = inner.indicators.lock().unwrap();
    let mut cache = inner.cache.lock().unwrap();
    if cache.len() != inds.len() {
        cache.resize_with(inds.len(), || None);
    }
    for i in 0..inds.len() {
        let due = force
            || match &cache[i] {
                None => true,
                Some(c) => now.saturating_sub(c.refreshed_at_ms) >= inds[i].policy.interval_ms,
            };
        if due {
            let component = inds[i].indicator.check();
            cache[i] = Some(Cached {
                component,
                refreshed_at_ms: now,
            });
        }
    }
    let all = build_view(inner, &inds, &cache, now, &stamp, None);
    let live = build_view(inner, &inds, &cache, now, &stamp, Some(ProbeKind::Liveness));
    let ready = build_view(
        inner,
        &inds,
        &cache,
        now,
        &stamp,
        Some(ProbeKind::Readiness),
    );
    let started = build_view(inner, &inds, &cache, now, &stamp, Some(ProbeKind::Startup));

    let overall = all.status;
    let mut current: BTreeMap<String, HealthStatus> = BTreeMap::new();
    for (i, reg) in inds.iter().enumerate() {
        let st = match &cache[i] {
            Some(c) => c.component.effective_status(),
            None => HealthStatus::Unknown,
        };
        current.insert(reg.indicator.name().to_string(), st);
    }
    drop(cache);
    drop(inds);
    *inner.snapshot.lock().unwrap() = Some(Arc::new(Snapshot {
        all,
        live,
        ready,
        started,
    }));

    emit_changes(inner, overall, current, &stamp);
}

/// Diff the new overall + per-component statuses against the previous rebuild and
/// emit a `subms-events` event for each transition. The first rebuild only sets
/// the baseline (no spurious events). The dispatch mode (sync/async) is the
/// dispatcher's; in async mode emit is a non-blocking enqueue.
fn emit_changes(
    inner: &Inner,
    overall: HealthStatus,
    current: BTreeMap<String, HealthStatus>,
    at: &str,
) {
    let mut changes: Vec<(String, HealthStatus, HealthStatus)> = Vec::new();
    {
        let mut prev = inner.prev_overall.lock().unwrap();
        if let Some(old) = *prev {
            if old != overall {
                changes.push(("overall".to_string(), old, overall));
            }
        }
        *prev = Some(overall);
    }
    {
        let mut prev = inner.prev_components.lock().unwrap();
        for (name, st) in &current {
            if let Some(old) = prev.get(name) {
                if old != st {
                    changes.push((name.clone(), *old, *st));
                }
            }
        }
        *prev = current;
    }
    for (scope, from, to) in changes {
        let event = Event::builder(HEALTH_STATUS_TOPIC)
            .level(level_for(to))
            .at(at)
            .message(&format!("{scope}: {} -> {}", from.as_str(), to.as_str()))
            .attr("scope", &scope)
            .attr("from", from.as_str())
            .attr("to", to.as_str())
            .build();
        inner.emitter.emit(event);
    }
}

struct Entry {
    status: HealthStatus,
    age_ms: u64,
    stale: bool,
    component: ComponentHealth,
}

fn build_view(
    inner: &Inner,
    inds: &[Registered],
    cache: &[Option<Cached>],
    now_ms: u64,
    refreshed_at: &str,
    filter: Option<ProbeKind>,
) -> View {
    let mut entries: BTreeMap<String, Entry> = BTreeMap::new();
    let mut contributed: Vec<HealthStatus> = Vec::new();
    for (i, reg) in inds.iter().enumerate() {
        if let Some(k) = filter {
            if !reg.policy.includes(k) {
                continue;
            }
        }
        let (component, refreshed_at_ms) = match &cache[i] {
            Some(c) => (c.component.clone(), c.refreshed_at_ms),
            None => (pending(), now_ms),
        };
        let eff = component.effective_status();
        // A non-critical failure never fails the report: it is demoted to Warn
        // (HTTP 200), so the pod keeps serving. Critical statuses pass through.
        let contrib =
            if !reg.policy.critical && matches!(eff, HealthStatus::Down | HealthStatus::Degraded) {
                HealthStatus::Warn
            } else {
                eff
            };
        contributed.push(contrib);
        let age_ms = now_ms.saturating_sub(refreshed_at_ms);
        let stale = (age_ms as f64) > (reg.policy.interval_ms as f64 * inner.stale_factor);
        entries.insert(
            reg.indicator.name().to_string(),
            Entry {
                status: eff,
                age_ms,
                stale,
                component,
            },
        );
    }
    let status = HealthStatus::aggregate(contributed);
    let json: Arc<str> = Arc::from(serialize_report(status, refreshed_at, &entries));
    View {
        status,
        code: code_for(status, filter),
        json,
    }
}

/// Probe-aware HTTP code. Liveness restarts the pod, so it only 503s on a hard
/// DOWN - a DEGRADED process must not be killed. Readiness 503s on DEGRADED/DOWN
/// (pull from rotation). Startup stays 503 until everything is UP/WARN. The full
/// `/health` document uses the readiness-style mapping ([`http_status_for`]).
fn code_for(status: HealthStatus, filter: Option<ProbeKind>) -> u16 {
    match filter {
        None => http_status_for(status),
        Some(ProbeKind::Liveness) => {
            if status == HealthStatus::Down {
                503
            } else {
                200
            }
        }
        Some(ProbeKind::Readiness) => {
            if matches!(status, HealthStatus::Degraded | HealthStatus::Down) {
                503
            } else {
                200
            }
        }
        Some(ProbeKind::Startup) => {
            if matches!(status, HealthStatus::Up | HealthStatus::Warn) {
                200
            } else {
                503
            }
        }
    }
}

fn pending() -> ComponentHealth {
    let mut c = ComponentHealth::unknown();
    c.details
        .insert("state".to_string(), JsonValue::Str("pending".to_string()));
    c
}

fn serialize_report(
    status: HealthStatus,
    refreshed_at: &str,
    entries: &BTreeMap<String, Entry>,
) -> String {
    let mut out = String::new();
    out.push('{');
    out.push_str("\"status\":");
    push_json_str(&mut out, status.as_str());
    out.push_str(",\"refreshed_at\":");
    push_json_str(&mut out, refreshed_at);
    if !entries.is_empty() {
        out.push_str(",\"components\":{");
        for (i, (name, e)) in entries.iter().enumerate() {
            if i > 0 {
                out.push(',');
            }
            push_json_str(&mut out, name);
            out.push(':');
            out.push('{');
            out.push_str("\"status\":");
            push_json_str(&mut out, e.status.as_str());
            out.push_str(&format!(",\"age_ms\":{}", e.age_ms));
            out.push_str(",\"stale\":");
            out.push_str(if e.stale { "true" } else { "false" });
            if !e.component.details.is_empty() {
                out.push_str(",\"details\":");
                push_map(&mut out, &e.component.details);
            }
            if !e.component.components.is_empty() {
                out.push_str(",\"components\":");
                push_component_map(&mut out, &e.component.components);
            }
            out.push('}');
        }
        out.push('}');
    }
    out.push('}');
    out
}

/// Built-in `server` indicator: pid, hostname, and process uptime.
pub struct ServerIndicator {
    started: Instant,
    pid: u32,
    hostname: String,
}

impl ServerIndicator {
    pub fn new() -> Self {
        let hostname = std::env::var("HOSTNAME")
            .or_else(|_| std::env::var("COMPUTERNAME"))
            .unwrap_or_else(|_| "unknown".to_string());
        Self {
            started: Instant::now(),
            pid: std::process::id(),
            hostname,
        }
    }
}

impl Default for ServerIndicator {
    fn default() -> Self {
        Self::new()
    }
}

impl HealthIndicator for ServerIndicator {
    fn name(&self) -> &str {
        "server"
    }
    fn check(&self) -> ComponentHealth {
        ComponentHealth::up()
            .with_detail("pid", self.pid as i64)
            .with_detail("hostname", self.hostname.as_str())
            .with_detail("uptime_ms", self.started.elapsed().as_millis() as i64)
    }
}
